apps/google-analytics-4/src/components/FormControlServiceAccountKey.tsx

```diff
@@ -0,0 +1,275 @@
+  currentServiceAccountKeyId: ServiceAccountKeyId | null;
+  className?: string;
+}
+
+class AssertionError extends Error {}
+
+const styles = {
+  credentialsInput: css({
+    fontFamily: 'monospace',
+  }),
+  credentialsNote: css({
+    marginBottom: tokens.spacingM,
+  }),
+  credentialsExplanation: css({
+    marginTop: tokens.spacingL,
+  }),
+};
+
+const placeholderText = `{
+  "type": "service_account",
+  ...
+}`;
+
+const keysOfServiceAccountKey: Array<keyof ServiceAccountKey> = [
+  'type',
+  'project_id',
+  'private_key_id',
+  'private_key',
+  'client_email',
+  'client_id',
+  'auth_uri',
+  'token_uri',
+  'auth_provider_x509_cert_url',
+  'client_x509_cert_url',
+];
+
+const convertKeyFileToServiceAccountKey = (keyFile: string): ServiceAccountKey => {
+  const parsedKeyFile = JSON.parse(keyFile);
+  assertServiceAccountKey(parsedKeyFile);
+
+  // ensure key file never contains extraneous keys
+  return pick(parsedKeyFile, keysOfServiceAccountKey);
+};
+
+const convertServiceAccountKeyToServiceAccountKeyId = (
+  serviceAccountKey: ServiceAccountKey
+): ServiceAccountKeyId => ({
+  id: serviceAccountKey.private_key_id,
+  clientEmail: serviceAccountKey.client_email,
+  clientId: serviceAccountKey.client_id,
+  projectId: serviceAccountKey.project_id,
+});
+
+function assertServiceAccountKey(value: any): asserts value is ServiceAccountKey {
+  if (value?.type !== 'service_account')
+    throw new AssertionError("Key file `type` must be 'service_account'");
+
+  const missingKeys = reject(keysOfServiceAccountKey, (key) => key in value);
+  if (missingKeys.length > 0)
+    throw new AssertionError(`Key file is missing the following keys: ${missingKeys.join(', ')}`);
+
+  const notStringValues = reject(keysOfServiceAccountKey, (key) => typeof value[key] === 'string');
+  if (notStringValues.length > 0)
+    throw new AssertionError(
+      `Key file has invalid values at the following keys: ${notStringValues.join(', ')}`
+    );
+}
+
+const FormControlServiceAccountKey = ({
+  setServiceAccountKey,
+  setServiceAccountKeyId,
+  setIsInvalid,
+  currentServiceAccountKeyId,
+  className,
+}: FormControlServideAccountKeyFileProps) => {
+  const [isExpanded, setIsExpanded] = useState<boolean>(false);
+  const [serviceAccountKeyFile, setServiceAccountKeyFile] = useState<string>('');
+  const [formControlState, setFormControlState] = useState<FormControlState>({
+    isInvalid: false,
+    errorMessage: '',
+  });
+
+  const setValidServiceAccountKey = useCallback(
+    (newServiceAccountKey: ServiceAccountKey | null) => {
+      setServiceAccountKey(newServiceAccountKey);
+      setServiceAccountKeyId(
+        newServiceAccountKey
+          ? convertServiceAccountKeyToServiceAccountKeyId(newServiceAccountKey)
+          : null
+      );
+      setFormControlState({
+        isInvalid: false,
+        errorMessage: '',
+      });
+    },
+    [setFormControlState, setServiceAccountKey, setServiceAccountKeyId]
```

Comment:
The main reason for this is feedback from the linter which centers around enforcing these principles: https://reactjs.org/docs/hooks-rules.html

Basically this is just a convenience helper function (since a few different states are set in concert when valid vs. invalid) and since these call the set state hooks, React asks to be explicit about declaring them within useCallback and being explicit about the dependencies OR just linlining the whole thing in useEffect. 

I believe I had it inlined before which worked fine, I find it just slightly harder to reason about inlined function definitions so I preferred defining this as a callback. But I'd be happy to go with whatever seems most obvious / unsurprising in your eyes.

Reply:
All good - it's better to have more than missing one. I would've assumed that linters don't complain here anymore. 🤷